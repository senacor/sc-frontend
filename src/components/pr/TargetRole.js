import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles/index';
import ListItem from '@material-ui/core/ListItem';
import StepSlider from './StepSlider';
import Grid from '@material-ui/core/Grid/index';
import objectGet from 'object-get';
import * as actions from '../../actions';
import { translateContent } from '../translate/Translate';
import Typography from '@material-ui/core/Typography';

const styles = theme => ({
  outerGrid: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'stretch'
  },
  ratingTargetRole: {
    color: theme.palette.primary['300'],
    justifyContent: 'flex-end'
  },
  targetRole: {
    color: theme.palette.primary['300'],
    alignItems: 'flex-end'
  },
  simpleBlack: {
    color: '#000000'
  }
});

const compareTargetRoles = (() => {
  const orderedTargetRoleNames = [
    'PLATTFORMGESTALTER',
    'IT_SOLUTION_LEADER',
    'TRANSFORMATION_MANAGER',
    'IT_LIEFERSTEUERER',
    'ARCHITECT',
    'TECHNICAL_EXPERT',
    'LEAD_DEVELOPER'
  ];

  return (objTargetRoleOne, objTargetRoleTwo) => {
    let priorityTargetRoleOne = orderedTargetRoleNames.indexOf(
      objTargetRoleOne.prTargetRoleName
    );
    let priorityTargetRoleTwo = orderedTargetRoleNames.indexOf(
      objTargetRoleTwo.prTargetRoleName
    );

    return priorityTargetRoleOne > priorityTargetRoleTwo ? 1 : -1;
  };
})();

class TargetRole extends Component {
  buildTargetRoleInformation = () => {
    let targetRoleInformationOfEmployee =
      objectGet(this.props, 'prActive.prTargetRoleSet') || [];
    targetRoleInformationOfEmployee.sort(compareTargetRoles);

    const { prActive, classes, prFinalized, isActionPerformer } = this.props;
    let isDisabled = !isActionPerformer;

    return targetRoleInformationOfEmployee.map(targetRole => {
      return (
        <Grid
          key={targetRole.prTargetRoleName + '_Container'}
          container
          className={classes.outerGrid}
        >
          <Grid
            item
            xs={8}
            sm={8}
            md={8}
            lg={8}
            xl={8}
            className={classes.targetRole}
          >
            <div className={classes.simpleBlack}>
              <Typography>
                {translateContent(targetRole.prTargetRoleName)}
              </Typography>
            </div>
          </Grid>

          <Grid
            item
            xs={4}
            sm={4}
            md={4}
            lg={4}
            xl={4}
            className={classes.ratingTargetRole}
            id={'target-role-' + targetRole.prTargetRoleName}
          >
            <StepSlider
              key={targetRole.prTargetRoleName}
              isDisabled={isDisabled || prFinalized}
              prActive={prActive}
              rating={targetRole.rating}
              targetRoleName={targetRole.prTargetRoleName}
            />
          </Grid>
        </Grid>
      );
    });
  };

  render() {
    return (
      <ListItem>
        <Grid container>{this.buildTargetRoleInformation()}</Grid>
      </ListItem>
    );
  }
}

export const StyledComponent = withStyles(styles)(TargetRole);
export default connect(
  state => ({
    prActive: state.prs[state.prDetailId]
  }),
  {
    changeRatingTargetRole: actions.changeRatingTargetRole
  }
)(StyledComponent);
