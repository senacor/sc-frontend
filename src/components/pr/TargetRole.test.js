import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import { StyledComponent as TargetRole } from './TargetRole';

const countTargetRoles = 7;
const prTargetRoleSet = [
  {
    id: 22,
    prTargetRoleName: 'PLATTFORMGESTALTER',
    rating: 2
  },
  {
    id: 27,
    prTargetRoleName: 'TRANSFORMATION_MANAGER',
    rating: 2
  },
  {
    id: 23,
    prTargetRoleName: 'ARCHITECT',
    rating: 2
  },
  {
    id: 26,
    prTargetRoleName: 'IT_LIEFERSTEUERER',
    rating: 2
  },
  {
    id: 24,
    prTargetRoleName: 'LEAD_DEVELOPER',
    rating: 2
  },
  {
    id: 25,
    prTargetRoleName: 'IT_SOLUTION_LEADER',
    rating: 2
  },
  {
    id: 28,
    prTargetRoleName: 'TECHNICAL_EXPERT',
    rating: 2
  }
];

describe('TargetRole Component', () => {
  let shallow = createShallow({ dive: true });
  it('should match snapshot', () => {
    let wrapper = shallow(
      <TargetRole
        prById={{}}
        prActive={{ prTargetRoleSet }}
        isDisabled={false}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should have expected number of target roles', () => {
    let wrapper = shallow(
      <TargetRole
        prById={{}}
        prActive={{ prTargetRoleSet }}
        isDisabled={false}
      />
    );

    expect(wrapper.find(ListItemText)).toHaveLength(countTargetRoles);
  });

  it('should show the target roles in the designated order', () => {
    let wrapper = shallow(
      <TargetRole
        prById={{}}
        prActive={{ prTargetRoleSet }}
        isDisabled={false}
      />
    );

    let actualRoles = wrapper
      .find(ListItemText)
      .map(role => role.get(0).props.secondary);
    expect(actualRoles).toEqual([
      'Plattformgestalter',
      'IT Solution Leader',
      'Transformation Manager',
      'IT Liefersteuerer',
      'Architekt',
      'Technischer Experte',
      'Lead Developer'
    ]);
  });
});
