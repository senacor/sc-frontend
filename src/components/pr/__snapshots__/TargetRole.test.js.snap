// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetRole Component should match snapshot 1`] = `
<WithStyles(ListItem)>
  <WithStyles(Grid)
    container={true}
  >
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="PLATTFORMGESTALTER_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            Plattformgestalter
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-PLATTFORMGESTALTER"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="PLATTFORMGESTALTER"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="PLATTFORMGESTALTER"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="IT_SOLUTION_LEADER_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            IT Solution Leader
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-IT_SOLUTION_LEADER"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="IT_SOLUTION_LEADER"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="IT_SOLUTION_LEADER"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="TRANSFORMATION_MANAGER_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            Transformation Manager
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-TRANSFORMATION_MANAGER"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="TRANSFORMATION_MANAGER"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="TRANSFORMATION_MANAGER"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="IT_LIEFERSTEUERER_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            IT Liefersteuerer
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-IT_LIEFERSTEUERER"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="IT_LIEFERSTEUERER"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="IT_LIEFERSTEUERER"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="ARCHITECT_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            Architekt
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-ARCHITECT"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="ARCHITECT"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="ARCHITECT"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="TECHNICAL_EXPERT_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            Technischer Experte
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-TECHNICAL_EXPERT"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="TECHNICAL_EXPERT"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="TECHNICAL_EXPERT"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      className="TargetRole-outerGrid-1"
      container={true}
      key="LEAD_DEVELOPER_Container"
    >
      <WithStyles(Grid)
        className="TargetRole-targetRole-3"
        item={true}
        lg={8}
        md={8}
        sm={8}
        xl={8}
        xs={8}
      >
        <div
          className="TargetRole-simpleBlack-4"
        >
          <WithStyles(Typography)>
            Lead Developer
          </WithStyles(Typography)>
        </div>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="TargetRole-ratingTargetRole-2"
        id="target-role-LEAD_DEVELOPER"
        item={true}
        lg={4}
        md={4}
        sm={4}
        xl={4}
        xs={4}
      >
        <Connect(WithStyles(StepSlider))
          isDisabled={true}
          key="LEAD_DEVELOPER"
          prActive={
            Object {
              "prTargetRoleSet": Array [
                Object {
                  "id": 22,
                  "prTargetRoleName": "PLATTFORMGESTALTER",
                  "rating": 2,
                },
                Object {
                  "id": 25,
                  "prTargetRoleName": "IT_SOLUTION_LEADER",
                  "rating": 2,
                },
                Object {
                  "id": 27,
                  "prTargetRoleName": "TRANSFORMATION_MANAGER",
                  "rating": 2,
                },
                Object {
                  "id": 26,
                  "prTargetRoleName": "IT_LIEFERSTEUERER",
                  "rating": 2,
                },
                Object {
                  "id": 23,
                  "prTargetRoleName": "ARCHITECT",
                  "rating": 2,
                },
                Object {
                  "id": 28,
                  "prTargetRoleName": "TECHNICAL_EXPERT",
                  "rating": 2,
                },
                Object {
                  "id": 24,
                  "prTargetRoleName": "LEAD_DEVELOPER",
                  "rating": 2,
                },
              ],
            }
          }
          rating={2}
          targetRoleName="LEAD_DEVELOPER"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(ListItem)>
`;
